[DOCS] Sizing shards rewrite (#120346)
elastic/elasticsearch: docs/reference/how-to/size-your-shards.asciidoc

[[size-your-shards]]
== Size your shards

Proper shard sizing is crucial for maintaining the performance and stability of an {es} cluster. _Oversharding_ occurs when data is distributed across an excessive number of shards, which can degrade search performance and make the cluster unstable. Conversely, very large shards may slow down search operations and prolong recovery times after failures. To strike the right balance, the <<shard-size-recommendation,general guidelines>> are to aim for shard sizes between 10GB and 50GB, keeping the per-shard document count below 200 million. To ensure that each node is working optimally, it's important to distribute shards evenly across nodes. Having uneven shard distribution will make some nodes work harder than others leading to performance degradation and potential instability. 

Despite these general guidelines, it is good to develop a tailored <<create-a-sharding-strategy, sharding strategy>> that considers your specific infrastructure, use case, and performance expectations.

[discrete]
[[what-is-a-shard]]

> === What is a shard?

Reviewer: Shouldn't this section come first? 

A more logical flow would be:

- What is it? (definition)
- Why does it matter? (importance of proper sizing)
- How do I use it? (strategy)

Author: Yeah that's a good idea. I was initially thinking that the primary purpose of this doc was about sizing specifically but I agree this would make more sense

Author: Let me know if you agree me changing the headers to all the same size (as opposed to "Size Your Shards" being largest)

[[size-your-shards]]
== Size your shards

Each index in {es} is divided into one or more shards, each of which may be
replicated across multiple nodes to protect against hardware failures. If you

> are using <<data-streams>> then each data stream is backed by a sequence of

Reviewer 2: > If you are using <<data-streams>> then each data stream is backed by a sequence of indices

This sentence is not in the new text. Can we insert it somehow somewhere?

Reviewer: @thekofimensah I would defer to @kingherc and restore the information requested

Reviewer 2: Well I think the purpose was to remind people that data streams have many indices and each index can have many shards.

Reviewer: Yes exactly 👍

Author: Got it

indices. There is a limit to the amount of data you can store on a single node
so you can increase the capacity of your cluster by adding nodes and increasing
the number of indices and shards to match. However, each index and shard has
some overhead and if you divide your data across too many shards then the
overhead can become overwhelming. A cluster with too many indices or shards is

> said to suffer from _oversharding_. An oversharded cluster will be less

Reviewer 2: > However, each index and shard has some overhead and if you divide your data across too many shards then the overhead can become overwhelming. A cluster with too many indices or shards is said to suffer from _oversharding_.

The new text does not say that each shard has some overhead. Can we insert it somehow somewhere?

Author: Committed a change to add this to the "What is a shard" section.

> Proper shard sizing is crucial for maintaining the performance and stability of an {es} cluster. _Oversharding_ occurs when data is distributed across an excessive number of shards, which can degrade search performance and make the cluster unstable. Conversely, very large shards may slow down search operations and prolong recovery times after failures. To strike the right balance, the <<shard-size-recommendation,general guidelines>> are to aim for shard sizes between 10GB and 50GB, keeping the per-shard document count below 200 million. To ensure that each node is working optimally, it's important to distribute shards evenly across nodes. Having uneven shard distribution will make some nodes work harder than others leading to performance degradation and potential instability. 

Reviewer 2: Regarding the last sentences on shard distribution, the "Elasticsearch automatically balances shards within a data tier" section later in this page says the shards are automatically balanced by ES. So I wonder if we need those sentences.

Author: I was looking at the first paragraph being a summary of the page, gathering the most important information right at the top. Therefore some repetition is expected, do you think that's reasonable?

> Despite these general guidelines, developing a tailored <<create-a-sharding-strategy, sharding strategy>> that considers your specific infrastructure, use case, and performance expectations is essential. Regular monitoring and adjustments based on real-world performance will help ensure optimal cluster health.

Reviewer 2: ```suggestion
Despite these general guidelines, it is good to develop a tailored <<create-a-sharding-strategy, sharding strategy>> that considers your specific infrastructure, use case, and performance expectations.
```

[discrete]
[[what-is-a-shard]]
=== What is a shard?

A shard is a basic unit of storage in {es}. Every index is divided into one or more shards to help distribute data and workload across nodes in a cluster. This division allows {es} to handle large datasets and perform operations like searches and indexing efficiently but not without cost. Each index and shard has some overhead and if you divide your data across too many shards then the overhead will degrade performance. Shards play several key roles in {es}:

* *Data Distribution:* Each shard contains a portion of the data from the index. When you add more nodes to your cluster, {es} will spread the shards across the nodes, balancing the workload between them.

> * *Replication:* Shards can have replicas which are copies of the original shard. Replicas ensure data availability and improve search performance by allowing multiple nodes to handle requests for that shard.

Reviewer 2: Too many replicas might also be an overkill, since the primary needs to replicate operations to the replicas. I wanted to say this in a comment as I'm not sure it's worth mentioning in the documentation as I'm not sure how much of a problem it can be / is.

Author: Potentially I could add it for the oversharding section by saying too many shards or replicas can lead to oversharding, what do you think?

> * *Parallel Processing:* Shards enable {es} to process queries in parallel across nodes, making searches faster and more efficient.

Reviewer 2: ```suggestion
* *Parallel Processing:* Shards enable {es} to distribute indexing of documents, and process queries in parallel across shards, making ingestion and searches faster and more efficient.
```

> By effectively using shards, {es} can scale horizontally and provide fault tolerance, ensuring your data is distributed and queries are processed efficiently.

Reviewer 2: ```suggestion
By effectively using shards, {es} can scale horizontally and provide fault tolerance, ensuring your data is distributed and indexing and searches are processed efficiently.
```

[[size-your-shards]]
=== Size your shards

> Proper shard sizing is crucial for maintaining the performance and stability of an {es} cluster. _Oversharding_ occurs when data is distributed across an excessive number of shards, which can degrade search performance and make the cluster unstable. Conversely, very large shards may slow down search operations and prolong recovery times after failures. To strike the right balance, the <<shard-size-recommendation,general guidelines>> are to aim for shard sizes between 10GB and 50GB, keeping the per-shard document count below 200 million. To ensure that each node is working optimally, it's important to distribute shards evenly across nodes. Having uneven shard distribution will make some nodes work harder than others leading to performance degradation and potential instability. 

Reviewer 2: Reading the last couple of lines makes me think the reader might try to manually distribute shards. Maybe we should reformulate them to something like:

```
To ensure that each node is working optimally, {es} gradually balances shards across the cluster, as an uneven shard distribution would make some nodes work harder than others leading to performance degradation and potential instability.
```

Author: I agree with you but I feel like your edit misses that they are trying to figure out how to decide how many shards to put per index and I want to avoid them having 4 shards on a 3 node cluster for example. 

What about:

To strike the right balance, the <<shard-size-recommendation,general guidelines>> are to aim for shard sizes between 10GB and 50GB, keeping the per-shard document count below 200 million. To ensure that each node is working optimally, it's important to distribute shards evenly across nodes. Uneven distribution can cause some nodes to work harder than others, leading to performance degradation and instability. While Elasticsearch automatically balances shards, it’s important to configure your indices with an appropriate number of shards and replicas to facilitate even distribution across nodes.

> If you are using <<data-streams>> then each data stream is backed by a sequence of indices which are all broken into their own shards. Proper allocation of shards for each backing index is crucial to maintain performance and stability. 

Reviewer 2: ```suggestion
If you are using <<data-streams>> then each data stream is backed by a sequence of indices, each index potentially having multiple shards. 
```

(deleting the last sentence since the distribution of shards was mentioned exactly above at the end of the previous paragraph)

* *Data Distribution:* Each shard contains a portion of the data from the index. When you add more nodes to your cluster, {es} will spread the shards across the nodes, balancing the workload between them.
* *Replication:* Shards can have replicas which are copies of the original shard. Replicas ensure data availability and improve search performance by allowing multiple nodes to handle requests for that shard.
* *Parallel Processing:* Shards enable {es} to distribute indexing of documents, and process queries in parallel across shards, making ingestion and searches faster and more efficient.

By effectively using shards, {es} can scale horizontally and provide fault tolerance, ensuring your data is distributed and indexing and searches are processed efficiently.

[discrete]
[[sizing-shard-guidelines]]
=== Sizing Shard Guidelines

> Proper shard sizing is crucial for maintaining the performance and stability of an {es} cluster. _Oversharding_ occurs when data is distributed across an excessive number of shards, which can degrade search performance and make the cluster unstable. Conversely, very large shards may slow down search operations and prolong recovery times after failures. 

Author: What do you think about 

Proper shard sizing is crucial for maintaining the performance and stability of an {es} cluster. _Oversharding_ occurs when data is distributed across an excessive number of shards, which can degrade search performance and make the cluster unstable. This can happen not only from having too many primary shards but also from an excessive number of replicas. Conversely, very large shards may slow down search operations and prolong recovery times after failures. 

@kingherc

Reviewer 2: Ah nice! Maybe simply

```
Proper shard sizing is crucial for maintaining the performance and stability of an {es} cluster. Oversharding occurs when data is distributed across an excessive number of shards (primary or replica), which can degrade search performance and make the cluster unstable. Conversely, very large shards may slow down search operations and prolong recovery times after failures.
```
